src/mame/ussr/juku.cpp

```diff
@@ -0,0 +1,849 @@
+		m_speaker->level_w(m_beep_state << (m_beep_level = BIT(data, 4)));
+
+	m_key_encoder->update();
+}
+
+uint8_t juku_state::pio0_portb_r()
+{
+	// 7-------  ctrl
+	// -6------  shift
+	// --54----  contrdat?
+	// ----321-  keyboard data
+	// -------0  key pressed
+
+	uint8_t data = 0;
+
+	data |= m_key_special->read();
+	data |= m_contrdat << 4;
+	data |= m_key_encoder->output_r() << 1;
+	data |= m_key_encoder->output_valid_r();
+
+	return data;
+}
+
+void juku_state::pio0_portc_w(uint8_t data)
+{
+	// 7-------  (cas?) pof
+	// -6------  (cas?) stop / floppy side select
+	// --5-----  (cas?) rn / floppy drive select
+	// ---4----  (cas?) ff / floppy density (sd = 1, dd = 0)
+	// ----3---  (cas?) play / floppy size (8" = 1, 5.25" = 0)
+	// -----2--  (cas?) rec / floppy motor (on = 1, off = 0)
+	// ------10  memory mode
+
+	floppy_image_device *floppy = m_floppy[BIT(data, 5)]->get_device();
+	m_fdc->set_floppy(floppy);
+	m_fdc->dden_w(BIT(data,4));
+	// TODO: 8" floppy select
+
+	if (floppy) {
+		floppy->mon_w(!BIT(data,2));
+		floppy->ss_w(BIT(data, 6));
+	}
+
+	m_bank->set_bank(data & 0x03);
+}
+
+void juku_state::machine_start()
+{
+	m_ram = std::make_unique<uint8_t []>(0x10000);
+
+	membank("rom_d800")->set_base(memregion("maincpu")->base() + 0x1800);
+
+	membank("ram_0000")->set_base(&m_ram[0x0000]);
+	membank("ram_4000")->set_base(&m_ram[0x4000]);
+	membank("ram_c000")->set_base(&m_ram[0xc000]);
+
+	// register for save states
+	save_item(NAME(m_width));
+	save_item(NAME(m_height));
+	save_item(NAME(m_hbporch));
+	save_item(NAME(m_vbporch));
+	save_item(NAME(m_contrdat));
+	save_item(NAME(m_monitor_bits));
+	save_item(NAME(m_height_lsb));
+	save_item(NAME(m_vblank_period_lsb));
+	save_item(NAME(m_empty_screen_on_update));
+	save_item(NAME(m_beep_state));
+	save_item(NAME(m_beep_level));
+	save_item(NAME(m_fdc_cur_cmd));
+	save_pointer(NAME(m_ram), 0x10000);
+}
+
+void juku_state::machine_reset()
+{
+	m_bank->set_bank(0);
+	m_key_encoder->enable_input_w(0);
+	m_beep_state = 0;
+	m_beep_level = 0;
+	m_contrdat = 0b11;
+	m_fdc_cur_cmd = 0;
+	m_width = DEFAULT_WIDTH, m_height = DEFAULT_HEIGHT;
+	m_hbporch = HORIZ_BACK_PORCH, m_vbporch = VERT_BACK_PORCH;
+	m_height_lsb = -1;
+	m_vblank_period_lsb = -1;
+	m_monitor_bits = 0U;
+	m_empty_screen_on_update = 0;
+}
+
+
+//**************************************************************************
+//  MACHINE DEFINTIONS
+//**************************************************************************
+
+void juku_state::juku(machine_config &config)
+{
+	// КР580ВМ80A @ 2 MHz (=KP580ИK80А)
+	I8080A(config, m_maincpu, 20_MHz_XTAL/10);
+	m_maincpu->set_addrmap(AS_PROGRAM, &juku_state::mem_map);
+	m_maincpu->set_addrmap(AS_IO, &juku_state::io_map);
+	m_maincpu->in_inta_func().set("pic", FUNC(pic8259_device::acknowledge));
+
+	ADDRESS_MAP_BANK(config, m_bank);
+	m_bank->set_map(&juku_state::bank_map);
+	m_bank->set_data_width(8);
+	m_bank->set_addr_width(18);
+	m_bank->set_stride(0x10000);
```

Comment:
This should really be converted to use `memory_view` rather than `address_map_bank_device`.